src/v/raft/seq_stm.cc

```diff
@@ -0,0 +1,348 @@
+              storage::snapshot_writer& writer) mutable {
+                return writer.write_metadata(std::move(data_size_buf))
+                  .then([&writer, data = std::move(data)]() mutable {
+                      return write_iobuf_to_output_stream(
+                        std::move(data), writer.output());
+                  })
+                  .finally([&writer] { return writer.close(); })
+                  .then([this, &writer] {
+                      return _snapshot_mgr.finish_snapshot(writer);
+                  });
+            });
+      });
+}
+
+ss::future<> seq_stm::make_snapshot() {
+    return _op_lock.with([this]() {
+        auto f = wait_for_snapshot_hydrated();
+        return f.then([this] { return do_make_snapshot(); });
+    });
+}
+
+ss::future<> seq_stm::ensure_snapshot_exists(model::offset target_offset) {
+    return _op_lock.with([this, target_offset]() {
+        auto f = wait_for_snapshot_hydrated();
+
+        return f.then([this, target_offset] {
+            if (target_offset <= _last_snapshot_offset) {
+                return ss::now();
+            }
+            return wait(target_offset, model::no_timeout)
+              .then([this, target_offset]() {
+                  vassert(
+                    target_offset <= _insync_offset,
+                    "after we waited for target_offset ({}) _insync_offset "
+                    "({}) should have matched it or bypassed",
+                    target_offset,
+                    _insync_offset);
+                  return do_make_snapshot();
+              });
+        });
+    });
+}
+
+ss::future<> seq_stm::do_make_snapshot() {
+    vlog(raftlog.trace, "saving seq_stm snapshot");
+    snapshot data;
+    data.offset = _insync_offset;
+    for (auto& entry : _seq_table) {
+        data.entries.push_back(entry.second);
+    }
+
+    iobuf v_buf;
+    reflection::adl<snapshot>{}.to(v_buf, data);
+    return persist_snapshot(std::move(v_buf))
+      .then([this, snapshot_offset = data.offset]() {
+          if (snapshot_offset >= _last_snapshot_offset) {
+              _last_snapshot_offset = snapshot_offset;
+          }
+          return _snapshot_mgr.remove_partial_snapshots();
+      });
+}
+
+ss::future<> seq_stm::start() {
+    _oldest_session = model::timestamp::now();
+    return _snapshot_mgr.open_snapshot().then(
+      [this](std::optional<storage::snapshot_reader> reader) {
+          auto f = ss::now();
+          if (reader) {
+              f = ss::do_with(
+                std::move(*reader), [this](storage::snapshot_reader& reader) {
+                    return hydrate_snapshot(reader).finally([this, &reader] {
+                        auto offset = std::max(
+                          _insync_offset, _c->start_offset());
+                        if (offset >= model::offset(0)) {
+                            set_next(offset);
+                        }
+                        _resolved_when_snapshot_hydrated.set_value();
```

Comment:
it looks like the only place where `_resolved_when_snapshot_hydrated` is set is here in `start()` and other parts of the interface wait for this to be set. generally there is a bit of an unwritten contract in the code base by which `start()` acts as a barrier such that the caller starting a service ensures no concurrent access to the service interface until `start()` resolves.

definitely reasonable for exceptions to that pattern, and I think the following in the comment in the header is alluding to the desire to have fast start-up:

>  To make the start up faster by reading the last snapshot and catching up with events written after the snapshot was made instead of reading the whole log from the beginning.

let's make sure that the comment portrays how the interface is intended to be used when it comes to `start()` sequencing with other parts of the api.

Reply:
We have the following workflow around partition.h/.cc

1. create consensus
2. create partition
3. pass consensus to partition
4. inside partition:
   - create stm(consensus)
   - pass stm to consensus via stm_manager
   - start consensus
   - start stm

We can't start stm before starting consensus but once consensus has started it may get a chance to invoke `make_snapshot` or `ensure_snapshot_exists` on stm before it's started. `wait_for_snapshot_hydrated` inside those methods protects from this possibility.

Reply:
add this as a comment to wait_for_snapshot_hydrated